File: ios/TeralFaceDetectorModule.swift
import ExpoModulesCore
import UIKit
import Vision

/**
 Detecta caras en una imagen y devuelve sus bounding boxes en coordenadas
 normalizadas (0..1) sobre la imagen ya orientada segun su EXIF.

 El modulo no toca la imagen: quien censura es la app, dibujando en Skia con
 esas coordenadas. Asi el usuario puede mover, borrar o añadir manchas antes de
 exportar, y el nativo se limita a lo unico que JS no puede hacer.

 La deteccion corre sobre una version reducida de la imagen (por defecto 1000px
 en el lado mayor): Vision encuentra las mismas caras y tarda una fraccion.
 */
public final class TeralFaceDetectorModule: Module {
  public func definition() -> ModuleDefinition {
    Name("TeralFaceDetector")

    AsyncFunction("detectFaces") { (uri: String, options: FaceDetectionOptions?) -> [String: Any] in
      return try self.detectFaces(uri: uri, options: options ?? FaceDetectionOptions())
    }

    AsyncFunction("detectTattoos") { (uri: String, options: TattooDetectionOptions?) -> [String: Any] in
      return try self.detectTattoos(uri: uri, options: options ?? TattooDetectionOptions())
    }

    Function("isTattooDetectionAvailable") {
      return (try? self.sharedTattooDetector()) != nil
    }
  }

  // MARK: - Tatuajes

  /// El modelo tarda en cargarse; se reutiliza entre llamadas.
  private var tattooDetector: TattooDetector?

  private func sharedTattooDetector() throws -> TattooDetector {
    if let existing = tattooDetector {
      return existing
    }
    let created = try TattooDetector()
    tattooDetector = created
    return created
  }

  private func detectTattoos(uri: String, options: TattooDetectionOptions) throws -> [String: Any] {
    let image = try loadImage(uri: uri)
    let detector = try sharedTattooDetector()

    let detections = try detector.detect(
      image: image,
      minConfidence: options.minConfidence,
      iouThreshold: options.iouThreshold
    )

    return [
      "imageWidth": Double(image.size.width * image.scale),
      "imageHeight": Double(image.size.height * image.scale),
      "tattoos": detections.map { detection in
        let clamped = self.clampToUnitSquare(
          CGRect(x: detection.x, y: detection.y, width: detection.width, height: detection.height)
        )
        return [
          "x": Double(clamped.origin.x),
          "y": Double(clamped.origin.y),
          "width": Double(clamped.width),
          "height": Double(clamped.height),
          "confidence": detection.confidence
        ]
      }
    ]
  }

  // MARK: - Deteccion

  private func detectFaces(uri: String, options: FaceDetectionOptions) throws -> [String: Any] {
    let image = try loadImage(uri: uri)

    // `size` de UIImage ya viene con la orientacion aplicada, asi que las
    // dimensiones que reportamos coinciden con lo que la app va a pintar.
    let imageWidth = image.size.width * image.scale
    let imageHeight = image.size.height * image.scale

    // El render normaliza la orientacion, por lo que Vision recibe siempre una
    // imagen `.up` y no hay que corregir ejes despues.
    guard let cgImage = uprightCGImage(from: image, maxSize: CGFloat(options.maxDetectionSize)) else {
      throw FaceDetectionError.imageRenderFailed(uri)
    }

    // Landmarks en vez de solo rectangulos: la misma pasada da la caja de la
    // cara y los puntos de los ojos, asi que la banda ocular no cuesta nada.
    let request = VNDetectFaceLandmarksRequest()
    let handler = VNImageRequestHandler(cgImage: cgImage, orientation: .up, options: [:])

    do {
      try handler.perform([request])
    } catch {
      throw FaceDetectionError.detectionFailed(error.localizedDescription)
    }

    let observations = request.results ?? []
    let faces = observations
      .filter { Double($0.confidence) >= options.minConfidence }
      .compactMap { observation -> [String: Any]? in
        // Vision usa origen abajo-izquierda; la app dibuja con origen arriba-izquierda.
        var faceRect = observation.boundingBox
        faceRect.origin.y = 1 - faceRect.origin.y - faceRect.height

        var rect = faceRect
        var angle: CGFloat = 0

        if options.region == .eyes {
          // Sin landmarks no hay banda que dibujar, y devolver la cara entera en
          // su lugar taparia mucho mas de lo que se pidio.
          guard let band = EyeBand.from(observation: observation, scale: CGFloat(options.eyeBandScale)) else {
            return nil
          }
          rect = band.rect
          angle = band.angle
        }

        let clamped = clampToUnitSquare(rect)

        return [
          "x": Double(clamped.origin.x),
          "y": Double(clamped.origin.y),
          "width": Double(clamped.width),
          "height": Double(clamped.height),
          "confidence": Double(observation.confidence),
          "region": options.region.rawValue,
          "angle": Double(angle)
        ]
      }

    return [
      "imageWidth": Double(imageWidth),
      "imageHeight": Double(imageHeight),
      "faces": faces
    ]
  }

  // MARK: - Carga de la imagen

  private func loadImage(uri: String) throws -> UIImage {
    let url: URL
    if uri.hasPrefix("/") {
      url = URL(fileURLWithPath: uri)
    } else if let parsed = URL(string: uri), parsed.scheme != nil {
      url = parsed
    } else {
      throw FaceDetectionError.unsupportedUri(uri)
    }

    guard let data = try? Data(contentsOf: url), let image = UIImage(data: data) else {
      throw FaceDetectionError.imageLoadFailed(uri)
    }

    return image
  }

  /**
   Devuelve la imagen redibujada con orientacion `.up` y reducida para que su
   lado mayor no pase de `maxSize`. Las imagenes mas pequeñas no se amplian.
   */
  private func uprightCGImage(from image: UIImage, maxSize: CGFloat) -> CGImage? {
    let pixelSize = CGSize(width: image.size.width * image.scale, height: image.size.height * image.scale)
    let longestSide = max(pixelSize.width, pixelSize.height)
    let scale = longestSide > maxSize && maxSize > 0 ? maxSize / longestSide : 1

    let targetSize = CGSize(width: (pixelSize.width * scale).rounded(), height: (pixelSize.height * scale).rounded())
    guard targetSize.width >= 1, targetSize.height >= 1 else { return nil }

    let format = UIGraphicsImageRendererFormat.default()
    format.scale = 1
    format.opaque = true

    let rendered = UIGraphicsImageRenderer(size: targetSize, format: format).image { _ in
      image.draw(in: CGRect(origin: .zero, size: targetSize))
    }

    return rendered.cgImage
  }

  private func clampToUnitSquare(_ rect: CGRect) -> CGRect {
    let minX = max(0, rect.minX)
    let minY = max(0, rect.minY)
    let maxX = min(1, rect.maxX)
    let maxY = min(1, rect.maxY)

    return CGRect(x: minX, y: minY, width: max(0, maxX - minX), height: max(0, maxY - minY))
  }
}

// MARK: - Errores

private enum FaceDetectionError: Error, LocalizedError {
  case unsupportedUri(String)
  case imageLoadFailed(String)
  case imageRenderFailed(String)
  case detectionFailed(String)

  var errorDescription: String? {
    switch self {
    case .unsupportedUri(let uri): return "URI no soportada: \(uri)"
    case .imageLoadFailed(let uri): return "No se pudo cargar la imagen: \(uri)"
    case .imageRenderFailed(let uri): return "No se pudo preparar la imagen para detectar: \(uri)"
    case .detectionFailed(let reason): return "Fallo la deteccion de caras: \(reason)"
    }
  }
}

File: android/src/main/java/ai/teral/facedetector/TeralFaceDetectorModule.kt
package ai.teral.facedetector

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.PointF
import android.media.FaceDetector
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/** Tope de la API nativa: pedir mas no aporta y reserva memoria de mas. */
private const val MAX_FACES = 20

/**
 * Ancho de la banda ocular como multiplo de la distancia entre ojos.
 *
 * Un ojo mide en torno a 0.35 de esa distancia, asi que cubrir ambos pide 1.35 y
 * el resto es margen. Con valores mayores la banda se come casi todo el ancho de
 * la cara y deja de tener sentido frente a taparla entera.
 */
private const val EYE_BAND_WIDTH_FACTOR = 1.7f

/** Alto de la banda ocular, en la misma unidad. Un ojo alto mide unos 0.2. */
private const val EYE_BAND_HEIGHT_FACTOR = 0.5f

/**
 * Detecta caras y devuelve sus bounding boxes en coordenadas normalizadas (0..1)
 * sobre la imagen ya orientada segun su EXIF.
 *
 * El modulo no modifica la imagen: quien censura es la app, dibujando en Skia con
 * esas coordenadas. Asi el usuario puede mover, borrar o añadir manchas antes de
 * exportar, y el nativo se limita a lo unico que JS no puede hacer.
 *
 * Usa `android.media.FaceDetector`, que es parte del sistema: funciona sin red,
 * sin Google Play Services y sin modelo empaquetado. A cambio detecta caras
 * frontales y poco mas; si hace falta precisión, este es el punto donde se
 * sustituye por ML Kit sin tocar la API.
 */
class TeralFaceDetectorModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw MissingContextException()

  override fun definition() = ModuleDefinition {
    Name("TeralFaceDetector")

    AsyncFunction("detectFaces") { uri: String, options: FaceDetectionOptions? ->
      detectFaces(uri, options ?: FaceDetectionOptions())
    }

    AsyncFunction("detectTattoos") { uri: String, options: TattooDetectionOptions? ->
      detectTattoos(uri, options ?: TattooDetectionOptions())
    }

    Function("isTattooDetectionAvailable") {
      runCatching { sharedTattooDetector() }.isSuccess
    }

    OnDestroy {
      tattooDetector?.close()
      tattooDetector = null
    }
  }

  // MARK: - Tatuajes

  /** Cargar el modelo cuesta; se reutiliza entre llamadas. */
  private var tattooDetector: TattooDetector? = null

  private fun sharedTattooDetector(): TattooDetector {
    return tattooDetector ?: TattooDetector(context).also { tattooDetector = it }
  }

  private fun detectTattoos(uri: String, options: TattooDetectionOptions): Map<String, Any> {
    val (imageWidth, imageHeight) = try {
      ImageLoader.readSize(context, uri)
    } catch (error: ImageLoadException) {
      throw TattooDetectionFailedException(error.message ?: "No se pudo leer la imagen")
    }

    val bitmap = try {
      ImageLoader.load(context, uri, options.maxDetectionSize)
    } catch (error: ImageLoadException) {
      throw TattooDetectionFailedException(error.message ?: "No se pudo cargar la imagen")
    }

    return try {
      val detections = sharedTattooDetector().detect(
        bitmap,
        options.minConfidence.toFloat(),
        options.iouThreshold.toFloat()
      )

      mapOf(
        "imageWidth" to imageWidth,
        "imageHeight" to imageHeight,
        "tattoos" to detections.map { detection ->
          val left = detection.x.coerceIn(0f, 1f)
          val top = detection.y.coerceIn(0f, 1f)
          mapOf(
            "x" to left,
            "y" to top,
            "width" to ((detection.x + detection.width).coerceIn(0f, 1f) - left).coerceAtLeast(0f),
            "height" to ((detection.y + detection.height).coerceIn(0f, 1f) - top).coerceAtLeast(0f),
            "confidence" to detection.confidence
          )
        }
      )
    } finally {
      bitmap.recycle()
    }
  }

  private fun detectFaces(uri: String, options: FaceDetectionOptions): Map<String, Any> {
    val (imageWidth, imageHeight) = try {
      ImageLoader.readSize(context, uri)
    } catch (error: ImageLoadException) {
      throw FaceDetectionFailedException(error.message ?: "No se pudo leer la imagen")
    }

    val source = try {
      ImageLoader.load(context, uri, options.maxDetectionSize)
    } catch (error: ImageLoadException) {
      throw FaceDetectionFailedException(error.message ?: "No se pudo cargar la imagen")
    }

    val detectable = toDetectableBitmap(source)

    return try {
      val detector = FaceDetector(detectable.width, detectable.height, MAX_FACES)
      val found = arrayOfNulls<FaceDetector.Face>(MAX_FACES)
      val count = detector.findFaces(detectable, found)

      val faces = (0 until count)
        .mapNotNull { found[it] }
        .filter { it.confidence() >= options.minConfidence }
        .map {
          toNormalizedRect(
            it,
            detectable.width.toFloat(),
            detectable.height.toFloat(),
            options.region,
            options.eyeBandScale.toFloat()
          )
        }

      mapOf(
        "imageWidth" to imageWidth,
        "imageHeight" to imageHeight,
        "faces" to faces
      )
    } finally {
      if (detectable != source) {
        detectable.recycle()
      }
      source.recycle()
    }
  }

  /**
   * `android.media.FaceDetector` solo acepta bitmaps RGB_565 de ancho par; con
   * cualquier otro formato devuelve cero caras sin avisar.
   */
  private fun toDetectableBitmap(source: Bitmap): Bitmap {
    val evenWidth = source.width and 0x1.inv()
    if (source.config == Bitmap.Config.RGB_565 && source.width == evenWidth) {
      return source
    }

    if (evenWidth <= 0 || source.height <= 0) {
      throw FaceDetectionFailedException("La imagen es demasiado pequeña para detectar caras")
    }

    val converted = Bitmap.createBitmap(evenWidth, source.height, Bitmap.Config.RGB_565)
    Canvas(converted).drawBitmap(source, 0f, 0f, null)
    return converted
  }

  /**
   * La API nativa no da un rectangulo de cara, sino el punto medio entre los ojos
   * y la distancia entre ellos. La banda ocular sale directamente de ahi; la cara
   * entera hay que construirla, con los factores que usa Signal (1.4 de ancho y
   * 0.4 de desplazamiento hacia abajo) para encuadrarla completa.
   */
  private fun toNormalizedRect(
    face: FaceDetector.Face,
    width: Float,
    height: Float,
    region: FaceRegion,
    eyeBandScale: Float
  ): Map<String, Any> {
    val midPoint = PointF().also { face.getMidPoint(it) }
    val eyesDistance = face.eyesDistance()

    val left: Float
    val top: Float
    val right: Float
    val bottom: Float

    if (region == FaceRegion.EYES) {
      val halfBandWidth = eyesDistance * EYE_BAND_WIDTH_FACTOR * eyeBandScale / 2f
      val halfBandHeight = eyesDistance * EYE_BAND_HEIGHT_FACTOR * eyeBandScale / 2f

      left = (midPoint.x - halfBandWidth) / width
      top = (midPoint.y - halfBandHeight) / height
      right = (midPoint.x + halfBandWidth) / width
      bottom = (midPoint.y + halfBandHeight) / height
    } else {
      val halfWidth = eyesDistance * 1.4f
      val yOffset = eyesDistance * 0.4f

      left = (midPoint.x - halfWidth) / width
      top = (midPoint.y - halfWidth + yOffset) / height
      right = (midPoint.x + halfWidth) / width
      bottom = (midPoint.y + halfWidth + yOffset) / height
    }

    val clampedLeft = left.coerceIn(0f, 1f)
    val clampedTop = top.coerceIn(0f, 1f)

    return mapOf(
      "x" to clampedLeft,
      "y" to clampedTop,
      "width" to (right.coerceIn(0f, 1f) - clampedLeft).coerceAtLeast(0f),
      "height" to (bottom.coerceIn(0f, 1f) - clampedTop).coerceAtLeast(0f),
      "confidence" to face.confidence(),
      "region" to region.value,
      // La API de deteccion de Android no reporta la pose de la cara, asi que la
      // banda va siempre horizontal. En iOS, donde hay landmarks, si se inclina.
      "angle" to 0f
    )
  }
}

internal class MissingContextException :
  CodedException("El contexto de React no esta disponible")

internal class FaceDetectionFailedException(reason: String) :
  CodedException("Fallo la deteccion de caras: $reason")

internal class TattooDetectionFailedException(reason: String) :
  CodedException("Fallo la deteccion de tatuajes: $reason")
